.devops/nix/package.nix

```diff
@@ -213,10 +213,16 @@ effectiveStdenv.mkDerivation (
         ;
 
       shell = mkShell {
+        NIX_LD_LIBRARY_PATH = lib.makeLibraryPath [
+          effectiveStdenv.cc.cc
+        ];
         name = "shell-${finalAttrs.finalPackage.name}";
         description = "contains numpy and sentencepiece";
         buildInputs = [ llama-python ];
         inputsFrom = [ finalAttrs.finalPackage ];
+        shellHook = ''
+          export LD_LIBRARY_PATH=$NIX_LD_LIBRARY_PATH
```

Comment:
- `LD_LIBRARY_PATH` makes `NIX_LD_LIBRARY_PATH` redundant
- Better to preprend to the (potentially) existing `LD_LIBRARY_PATH` because the variable can be used in two different ways: to override a service (which is what we do here), and to provide missing services (e.g. the impurely linked userspace drivers)

Reply:
On the second thought, there's no reason to prepend either. As long as `venv` doesn't set `LD_LIBRARY_PATH`, we're happy just appending:

```
shellHook = ''
  addToSearchPath "LD_LIBRARY_PATH" "${lib.getLib stdenv.cc.cc}/lib"
'';
```

or something of the sort.

Furthermore! We might want to set `NIX_LD` as per https://github.com/Mic92/nix-ld/tree/main#usage to ensure that native extensions from pypi use the same `ld.so` (in addition to `libc` and `libstdc++`) as our project.

I'm still a bit uncomfortable setting the global `LD_LIBRARY_PATH`... maybe we should tell NixOS users use `nix-ld` instead, because they should? Outside NixOS/in FHS environments there's probably no need for these overrides either

Reply:
Also if we set any of `NIX_LD_*` we should add a link to https://github.com/Mic92/nix-ld/tree/main#usage in the comments

Reply:
> To clarify, you were running these examples on NixOS?

yes, and that's where I encountered the `ImportError: libstdc++.so.6: cannot open shared object file: No such file or directory` error I put in the commit msg.

> I'm still a bit uncomfortable setting the global `LD_LIBRARY_PATH` unconditionally... maybe we should tell NixOS users use `nix-ld` instead, because they should? Outside NixOS/in FHS environments there's probably no need for these overrides either

hmm yes that's reasonable. although, `LD_LIBRARY_PATH` would only be set for the individual `nix-shell`/`nix develop` shell, and wouldn't therefore "globally contaminate" other shells.

> We might want to set NIX_LD as per https://github.com/Mic92/nix-ld/tree/main#usage to ensure that native extensions from pypi use the same ld.so (in addition to libc and libstdc++) as our project.

this seems like a valid concern which could motivate using `NIX_LD*`. being a Python-avoider, I do not know much about how that works, tho... 

personally, my bias would be towards requiring less of the user, and not forcing them to set up `nix-ld`. I would probably not want to do that and would just locally hack around the issue "within the context of the llama.cpp repo" (e.g. by setting `LD_LIBRARY_PATH`) rather than having to modify my NixOS system. I understand that not being appealing or "clean", as an official solution, tho.

Reply:
I adopted your suggestion in https://github.com/ggerganov/llama.cpp/pull/5078/commits/0360217e8fd3953f647788ef9774a0e8c43b50af

Reply:
>     LD_LIBRARY_PATH would only be set for the individual nix-shell/nix develop shell, and wouldn't therefore "globally contaminate" other shells.

By "global" I meant it doesn't just affect the venv's python, it also affects nixpkgs' python and generally all processes run from the shell. E.g. if one loads `/run/opengl-driver/lib/libcuda.so` inside this shell, it'll end up using the shell's libstdc++, even if the shell uses an older Nixpkgs than NixOS. With `nix-ld` there's slightly more isolation (in principle, still an issue, cf the readme for nix-ld) which is why on NixOS you're better of using that than setting `LD_LIBRARY_PATH` directly

Reply:
thx, that's a good point.